server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java

```diff
@@ -761,4 +823,80 @@ public long nextValue() throws IOException {
             }
         }
     }
+
+    /**
+     * Hybrid Collector that starts with OrdinalsCollector and switches to DirectCollector
+     * when memory consumption exceeds a threshold.
+     */
+    static class HybridCollector extends Collector {
+        private final HyperLogLogPlusPlus counts;
+        private final MurmurHash3Values hashValues;
+        private final long memoryThreshold;
+
+        private Collector activeCollector;
+        private final OrdinalsCollector ordinalsCollector;
+
+        HybridCollector(
+            HyperLogLogPlusPlus counts,
+            SortedSetDocValues ordinalValues,
+            MurmurHash3Values hashValues,
+            BigArrays bigArrays,
+            long memoryThreshold
+        ) {
+            this.counts = counts;
+            this.hashValues = hashValues;
+            this.memoryThreshold = memoryThreshold;
+
+            // Start with OrdinalsCollector
+            this.ordinalsCollector = new OrdinalsCollector(
+                counts,
+                ordinalValues,
+                bigArrays,
+                memoryThreshold,
+                Optional.of(this::handleMemoryLimitReached)
+            );
+            this.activeCollector = ordinalsCollector;
+        }
+
+        @Override
+        public void collect(int doc, long bucketOrd) throws IOException {
+            activeCollector.collect(doc, bucketOrd);
```

Comment:
This call should not be polymorphic and run into virtual call overhead, since only 2 implementations are accessible here. Although, the caller for `HybridCollector#collect` might already be seeing that overhead, which I guess is same earlier and now